fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -447,11 +447,16 @@ public RecordStoreState getRecordStoreState() {
     }
 
     private CompletableFuture<RecordStoreState> getRecordStoreStateAsync() {
-        if (recordStoreStateRef.get() != null) {
-            return CompletableFuture.completedFuture(recordStoreStateRef.get());
+        final MutableRecordStoreState localStoreState = recordStoreStateRef.get();
+        if (localStoreState != null) {
+            return CompletableFuture.completedFuture(localStoreState);
         }
+        beginRecordStoreStateRead();
         return preloadRecordStoreStateAsync()
-                .thenApply(ignore -> recordStoreStateRef.get());
+                .thenApply(ignore -> {
+                    endRecordStoreStateRead();
```

Comment:
`endRecordStoreStateRead()` needs to be called even if the futures fail, so it needs to be in a `whenComplete`.